frontend/apps/reader/modules/readerhighlight.lua

```diff
@@ -1611,6 +1613,97 @@ function ReaderHighlight:getHighlightBookmarkItem()
     end
 end
 
+function ReaderHighlight:_getInsertionPositionForRolling(page, highlight)
+    local position
+    local highlight_beginning = highlight.pos0
+    local highlight_end = highlight.pos1
+    local page_highlights = self.view.highlight.saved[page]
+    for i, cur_highlight in ipairs(page_highlights) do
+        local order = self.ui.document:compareXPointers(highlight_beginning, cur_highlight.pos0)
+        if order == 0 then
```

Comment:
You cannot compare nil with number.

Reply:
By comparing are you meaning line 1623? I can surely check if the order is explicitly a number, right?

I'm taking the `nil` value into consideration too. But, unlike this:
https://github.com/koreader/koreader/blob/82feb21684bfa98c795bd80b19f97a535524db6b/frontend/apps/reader/modules/readerbookmark.lua#L234-L244

I'm checking if the `order` value is exactly zero and then descending down to reassign the `order` value again by comparison.
Then I'm checking for `nil` explicitly so if there's a nil value from any level of the comparison, the loop breaks.

Reply:
If order is nil, line 1623 will give a crash.

Reply:
I've run this script:

```lua
order = nil
if order == 0 then
  print("it's zero")
end
if order == nil then
  print("it's nil")
end
```
and it is printing "it's nil". Am I missing something? (Lua version 5.4.4)

Reply:
https://github.com/koreader/koreader/pull/8616#issuecomment-1015443117

Reply:
It won't crash on line 1623. `nil == 0` is a perfectly valid comparison. `nil > 0` or `nil < 0` would've thrown an error. I have such a comparison at line 1629 however I have got a nil check in line 1626-1628 which breaks the loop so line 1629 will never produce a loop.

Reply:
In #8916 `return compare_xp < 0` was throwing the error since there weren't any checking for nil before that. `compare_xp == 0` was fine.

Reply:
Okay

Reply:
Should I mark this thread as resolved?